File: shared/src/commonMain/kotlin/energy/ftw/identity/Enrollment.kt
package energy.ftw.identity

import energy.ftw.APP_HOST
import energy.ftw.ENROLLMENT_PATH
import energy.ftw.isDevHost

/**
 * The enrollment payload.
 *
 *   https://app.ftw.energy/p#v2.<box_noise_pub>.<pairing_code>.<lan_hint>.<rendezvous_secret>
 *
 * Everything after the `#` is a URL fragment and is never sent in an HTTP
 * request. The trust anchor reaches the app optically.
 */
const val ENROLLMENT_VERSION = "v2"
const val BOX_KEY_BYTES = 32
const val PAIRING_CODE_BYTES = 16
const val RENDEZVOUS_SECRET_BYTES = 32
const val MAX_LAN_HINT_CHARS = 64

data class Enrollment(
    val boxStaticPublic: ByteArray,
    val pairingCode: ByteArray,
    val lanHint: String,
    val rendezvousSecret: ByteArray,
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Enrollment) return false
        return boxStaticPublic.contentEquals(other.boxStaticPublic) &&
            pairingCode.contentEquals(other.pairingCode) &&
            lanHint == other.lanHint &&
            rendezvousSecret.contentEquals(other.rendezvousSecret)
    }

    override fun hashCode(): Int {
        var h = boxStaticPublic.contentHashCode()
        h = 31 * h + pairingCode.contentHashCode()
        h = 31 * h + lanHint.hashCode()
        h = 31 * h + rendezvousSecret.contentHashCode()
        return h
    }
}

enum class EnrollmentErrorCode {
    E_QR_NOT_FTW,
    E_QR_VERSION,
    E_QR_SHAPE,
    E_QR_ENCODING,
    E_QR_KEY,
    E_QR_CODE,
    E_QR_HINT,
    E_QR_SECRET,
}

class EnrollmentError(
    message: String,
    val code: EnrollmentErrorCode,
    /** What the user does now. Never what broke inside. */
    val help: String,
) : Exception(message)

private const val SCAN_AGAIN =
    "That code did not read cleanly. Hold the phone steady and scan it again."
private const val WRONG_CODE =
    "That is not an FTW pairing code. Open your box's local dashboard, then Settings → FTW app → Show pairing code, and scan the QR shown there."
private const val APP_TOO_OLD =
    "This box needs a newer version of the app. Update the app, then scan again."
private const val BOX_TOO_OLD =
    "This box needs a software update before it can pair. Update the box, then scan again."

private val OUR_VERSION = ENROLLMENT_VERSION.removePrefix("v").toInt()

fun parseEnrollmentUrl(raw: String): Enrollment {
    val url = try {
        ParsedUrl.parse(raw)
    } catch (_: IllegalArgumentException) {
        throw EnrollmentError("not a URL: ${raw.take(24)}", EnrollmentErrorCode.E_QR_NOT_FTW, WRONG_CODE)
    }

    val httpsOrDev = url.scheme == "https" || (url.scheme == "http" && isDevHost(url.host))
    if (!httpsOrDev) {
        throw EnrollmentError("scheme ${url.scheme} is not https", EnrollmentErrorCode.E_QR_NOT_FTW, WRONG_CODE)
    }
    if (url.host != APP_HOST && !isDevHost(url.host)) {
        throw EnrollmentError("host ${url.host} is not FTW", EnrollmentErrorCode.E_QR_NOT_FTW, WRONG_CODE)
    }
    if (url.path != ENROLLMENT_PATH) {
        throw EnrollmentError(
            "path ${url.path} is not $ENROLLMENT_PATH",
            EnrollmentErrorCode.E_QR_NOT_FTW,
            WRONG_CODE,
        )
    }
    return parseEnrollmentFragment(url.fragment)
}

fun parseEnrollmentFragment(fragment: String): Enrollment {
    val body = if (fragment.startsWith('#')) fragment.substring(1) else fragment
    if (body.isEmpty()) {
        throw EnrollmentError("empty fragment", EnrollmentErrorCode.E_QR_NOT_FTW, WRONG_CODE)
    }

    val parts = body.split('.')
    val version = parts[0]

    if (version != ENROLLMENT_VERSION) {
        if (version.matches(Regex("^v\\d+$"))) {
            val n = version.substring(1).toInt()
            val help = if (n < OUR_VERSION) BOX_TOO_OLD else APP_TOO_OLD
            throw EnrollmentError("payload version $version", EnrollmentErrorCode.E_QR_VERSION, help)
        }
        throw EnrollmentError("fragment does not start with a version", EnrollmentErrorCode.E_QR_NOT_FTW, WRONG_CODE)
    }

    if (parts.size != 5) {
        throw EnrollmentError("${parts.size} segments, expected 5", EnrollmentErrorCode.E_QR_SHAPE, SCAN_AGAIN)
    }

    val boxStaticPublic = decodeSegment(parts[1], "box key")
    val pairingCode = decodeSegment(parts[2], "pairing code")
    val lanHint = decodeSegment(parts[3], "lan hint").decodeToString()
    val rendezvousSecret = decodeSegment(parts[4], "rendezvous secret")

    if (boxStaticPublic.size != BOX_KEY_BYTES) {
        throw EnrollmentError(
            "box key is ${boxStaticPublic.size} bytes, expected $BOX_KEY_BYTES",
            EnrollmentErrorCode.E_QR_KEY,
            SCAN_AGAIN,
        )
    }
    if (pairingCode.size != PAIRING_CODE_BYTES) {
        throw EnrollmentError(
            "pairing code is ${pairingCode.size} bytes, expected $PAIRING_CODE_BYTES",
            EnrollmentErrorCode.E_QR_CODE,
            SCAN_AGAIN,
        )
    }
    if (lanHint.length > MAX_LAN_HINT_CHARS || lanHint.any { it.code < 0x21 || it.code > 0x7e }) {
        throw EnrollmentError("lan hint is not a plain address", EnrollmentErrorCode.E_QR_HINT, SCAN_AGAIN)
    }
    if (rendezvousSecret.size != RENDEZVOUS_SECRET_BYTES) {
        throw EnrollmentError(
            "rendezvous secret is ${rendezvousSecret.size} bytes, expected $RENDEZVOUS_SECRET_BYTES",
            EnrollmentErrorCode.E_QR_SECRET,
            SCAN_AGAIN,
        )
    }

    return Enrollment(boxStaticPublic, pairingCode, lanHint, rendezvousSecret)
}

fun buildEnrollmentUrl(enrollment: Enrollment, origin: String = "https://$APP_HOST"): String {
    val fragment = listOf(
        ENROLLMENT_VERSION,
        encodeBase64url(enrollment.boxStaticPublic),
        encodeBase64url(enrollment.pairingCode),
        encodeBase64url(enrollment.lanHint.encodeToByteArray()),
        encodeBase64url(enrollment.rendezvousSecret),
    ).joinToString(".")
    return "$origin$ENROLLMENT_PATH#$fragment"
}

private fun decodeSegment(segment: String, what: String): ByteArray {
    try {
        return decodeBase64url(segment)
    } catch (err: Base64urlError) {
        throw EnrollmentError("$what: ${err.message}", EnrollmentErrorCode.E_QR_ENCODING, SCAN_AGAIN)
    }
}

/**
 * Tiny URL parser so commonMain does not depend on java.net.URI.
 * Handles https://host[:port]/path#fragment and http://localhost.
 */
internal data class ParsedUrl(
    val scheme: String,
    val host: String,
    val path: String,
    val fragment: String,
) {
    companion object {
        fun parse(raw: String): ParsedUrl {
            val schemeEnd = raw.indexOf("://")
            if (schemeEnd <= 0) throw IllegalArgumentException("no scheme")
            val scheme = raw.substring(0, schemeEnd)
            val rest = raw.substring(schemeEnd + 3)
            val hash = rest.indexOf('#')
            val beforeHash = if (hash >= 0) rest.substring(0, hash) else rest
            val fragment = if (hash >= 0) rest.substring(hash + 1) else ""
            val slash = beforeHash.indexOf('/')
            val hostPort = if (slash >= 0) beforeHash.substring(0, slash) else beforeHash
            val path = if (slash >= 0) beforeHash.substring(slash) else "/"
            if (hostPort.isEmpty()) throw IllegalArgumentException("no host")
            val host = hostPort.substringBefore(':')
            return ParsedUrl(scheme, host, path, fragment)
        }
    }
}

File: appleApp/FTWKit/Sources/FTWKit/Identity/Enrollment.swift
import Foundation

/// The pairing payload.
///
///     https://app.ftw.energy/p#v2.<box_noise_pub>.<pairing_code>.<lan_hint>.<rendezvous_secret>
///
/// Everything after `#` is a fragment and never sent in a request, so the
/// trust anchor reaches the phone optically and reaches no server on the way.
public struct Enrollment: Equatable, Sendable {
    /// The box's Noise static public key. The trust anchor.
    public var boxStaticPublic: Bytes
    /// Single use. The box refuses it a second time. Never log this.
    public var pairingCode: Bytes
    /// Where the box believes it can be reached on the LAN. May be empty.
    public var lanHint: String
    /// The long-lived secret the rotating relay handle is derived from.
    public var rendezvousSecret: Bytes

    public static let version = "v2"
    public static let host = Origin.appHost
    public static let path = "/p"
    public static let boxKeyBytes = 32
    public static let pairingCodeBytes = 16
    public static let rendezvousSecretBytes = 32
    public static let maxLanHintChars = 64

    public init(boxStaticPublic: Bytes, pairingCode: Bytes, lanHint: String, rendezvousSecret: Bytes) {
        self.boxStaticPublic = boxStaticPublic
        self.pairingCode = pairingCode
        self.lanHint = lanHint
        self.rendezvousSecret = rendezvousSecret
    }
}

/// What the user does next. Never which check failed.
public struct EnrollmentError: Error, Equatable, HelpfulError {
    public let code: String
    public let message: String
    public let help: String

    static let scanAgain = "That code did not read cleanly. Hold the phone steady and scan it again."
    static let wrongCode = "That is not an FTW pairing code. Open your box's local dashboard, then Settings → FTW app → Show pairing code, and scan the QR shown there."
    static let appTooOld = "This box needs a newer version of the app. Update the app, then scan again."
    static let boxTooOld = "This box needs a software update before it can pair. Update the box, then scan again."
}

/// An error that carries a sentence the person holding the phone can act on.
public protocol HelpfulError: Error {
    var help: String { get }
}

extension Enrollment {
    /// Parse whatever a camera, a paste or a link handed over: a full URL or
    /// just the fragment.
    public static func parse(scanned raw: String) throws -> Enrollment {
        let text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if text.isEmpty {
            throw EnrollmentError(code: "E_QR_NOT_FTW", message: "nothing scanned", help: "That code did not scan. Try again.")
        }
        return text.hasPrefix("#") ? try parse(fragment: text) : try parse(url: text)
    }

    public static func parse(url raw: String) throws -> Enrollment {
        guard let components = URLComponents(string: raw), let scheme = components.scheme else {
            throw EnrollmentError(code: "E_QR_NOT_FTW", message: "not a URL", help: EnrollmentError.wrongCode)
        }
        guard scheme.lowercased() == "https" else {
            throw EnrollmentError(code: "E_QR_NOT_FTW", message: "scheme \(scheme) is not https", help: EnrollmentError.wrongCode)
        }
        guard components.host?.lowercased() == host, components.port == nil, components.user == nil else {
            throw EnrollmentError(code: "E_QR_NOT_FTW", message: "host is not FTW", help: EnrollmentError.wrongCode)
        }
        guard components.path == path else {
            throw EnrollmentError(code: "E_QR_NOT_FTW", message: "path \(components.path) is not \(path)", help: EnrollmentError.wrongCode)
        }
        // The fragment as written, not percent-decoded: every segment is
        // base64url and a decoded one would be a different string.
        guard let hashIndex = raw.firstIndex(of: "#") else {
            throw EnrollmentError(code: "E_QR_NOT_FTW", message: "no fragment", help: EnrollmentError.wrongCode)
        }
        return try parse(fragment: String(raw[hashIndex...]))
    }

    public static func parse(fragment raw: String) throws -> Enrollment {
        let body = raw.hasPrefix("#") ? String(raw.dropFirst()) : raw
        if body.isEmpty {
            throw EnrollmentError(code: "E_QR_NOT_FTW", message: "empty fragment", help: EnrollmentError.wrongCode)
        }
        let parts = body.split(separator: ".", omittingEmptySubsequences: false).map(String.init)
        let v = parts[0]

        // Version before shape, and which side is behind decides the
        // sentence. Telling someone to update the wrong thing is worse than
        // saying nothing.
        if v != version {
            if v.count > 1, v.hasPrefix("v"), let n = Int(v.dropFirst()), v.dropFirst().allSatisfy(\.isNumber) {
                let ours = Int(version.dropFirst())!
                throw EnrollmentError(code: "E_QR_VERSION", message: "payload version \(v)", help: n < ours ? EnrollmentError.boxTooOld : EnrollmentError.appTooOld)
            }
            throw EnrollmentError(code: "E_QR_NOT_FTW", message: "fragment does not start with a version", help: EnrollmentError.wrongCode)
        }
        guard parts.count == 5 else {
            throw EnrollmentError(code: "E_QR_SHAPE", message: "\(parts.count) segments, expected 5", help: EnrollmentError.scanAgain)
        }

        let box = try segment(parts[1])
        let code = try segment(parts[2])
        let hintBytes = try segment(parts[3])
        let secret = try segment(parts[4])

        guard box.count == boxKeyBytes else {
            throw EnrollmentError(code: "E_QR_KEY", message: "box key is \(box.count) bytes", help: EnrollmentError.scanAgain)
        }
        guard code.count == pairingCodeBytes else {
            throw EnrollmentError(code: "E_QR_CODE", message: "pairing code is \(code.count) bytes", help: EnrollmentError.scanAgain)
        }
        // The hint becomes a connection target later, so it is checked here.
        guard let hint = String(bytes: hintBytes, encoding: .utf8),
              hint.count <= maxLanHintChars,
              hint.unicodeScalars.allSatisfy({ $0.value >= 0x21 && $0.value <= 0x7e }) else {
            throw EnrollmentError(code: "E_QR_HINT", message: "lan hint is not a plain address", help: EnrollmentError.scanAgain)
        }
        // Downstream this is HKDF input keying material, and a short secret
        // there is a handle an attacker can enumerate.
        guard secret.count == rendezvousSecretBytes else {
            throw EnrollmentError(code: "E_QR_SECRET", message: "rendezvous secret is \(secret.count) bytes", help: EnrollmentError.scanAgain)
        }
        return Enrollment(boxStaticPublic: box, pairingCode: code, lanHint: hint, rendezvousSecret: secret)
    }

    /// The inverse, for tests and the loopback box.
    public func url() -> String {
        let fragment = [
            Enrollment.version,
            Base64url.encode(boxStaticPublic),
            Base64url.encode(pairingCode),
            Base64url.encode(Array(lanHint.utf8)),
            Base64url.encode(rendezvousSecret),
        ].joined(separator: ".")
        return "https://\(Enrollment.host)\(Enrollment.path)#\(fragment)"
    }

    private static func segment(_ s: String) throws -> Bytes {
        do {
            return try Base64url.decode(s)
        } catch {
            throw EnrollmentError(code: "E_QR_ENCODING", message: "segment is not base64url", help: EnrollmentError.scanAgain)
        }
    }
}
